dev-tools/xtask/src/download.rs

```diff
@@ -0,0 +1,982 @@
+    #[clap(long, default_value = "out")]
+    output_dir: Utf8PathBuf,
+
+    /// The path to the versions and checksums directory.
+    #[clap(long, default_value = "tools")]
+    versions_dir: Utf8PathBuf,
+}
+
+pub async fn run_cmd(args: DownloadArgs) -> Result<()> {
+    let mut targets = BTreeSet::new();
+
+    for target in args.targets {
+        match target {
+            Target::All => {
+                // Add all targets, then remove the "All" variant because that
+                // isn't a real thing we can download.
+                let mut all = BTreeSet::from_iter(Target::iter());
+                all.remove(&Target::All);
+                targets.append(&mut all);
+            }
+            _ => _ = targets.insert(target),
+        }
+    }
+
+    let decorator = slog_term::TermDecorator::new().build();
+    let drain = slog_term::FullFormat::new(decorator).build().fuse();
+    let drain = slog_async::Async::new(drain).build().fuse();
+    let log = Logger::root(drain, o!());
+
+    let mut all_downloads = targets
+        .iter()
+        .map(|target| {
+            let log = log.clone();
+            let output_dir = args.output_dir.clone();
+            let versions_dir = args.versions_dir.clone();
+            async move {
+                info!(&log, "Download target: {target:?}");
+
+                let downloader = Downloader::new(
+                    log.new(o!("target" => format!("{target:?}"))),
+                    &output_dir,
+                    &versions_dir,
+                );
+
+                match target {
+                    Target::All => {
+                        bail!("We should have already filtered this 'All' target out?");
+                    }
+                    Target::Clickhouse => downloader.download_clickhouse().await,
+                    Target::Cockroach => downloader.download_cockroach().await,
+                    Target::Console => downloader.download_console().await,
+                    Target::DendriteOpenapi => {
+                        downloader.download_dendrite_openapi().await
+                    }
+                    Target::DendriteStub => downloader.download_dendrite_stub().await,
+                    Target::MaghemiteMgd => downloader.download_maghemite_mgd().await,
+                    Target::MaghemiteOpenapi => {
+                        downloader.download_maghemite_openapi().await
+                    }
+                    Target::Softnpu => downloader.download_softnpu().await,
+                    Target::Thundermuffin => {
+                        downloader.download_thundermuffin().await
+                    }
+                    Target::TransceiverControl => {
+                        downloader.download_transceiver_control().await
+                    }
+                }.context("Failed to download {target:?}")?;
+                Ok(())
+            }
+        })
+        .collect::<futures::stream::FuturesUnordered<_>>();
```

Comment:
If we're downloading multiple things, might we want to `tokio::spawn` the `download_${FOO}` calls, and put their `JoinHandle`s in the `FuturesUnordered`, instead? That way, the various downloads may run in parallel on multiple CPU cores.

Reply:
Great advice, done in 3144d5d02e93379cc084ddc6854282fea3e2103b